src/aspire/source/micrograph.py

```diff
@@ -1,16 +1,223 @@
+
+    def _images_from_array(self, indices):
+        """
+        Accesses and returns micrographs from a Numpy array.
+
+        :param indices: A 1-D Numpy array of integer indices.
+        :return: An array backed `MicrographSource` object representing the micrographs for `indices`.
+        """
+        return MicrographSource(self._data[indices])
+
+    def _images_from_files(self, indices):
+        """
+        Accesses and returns micrographs from disk.
+
+        :param indices: A 1-D Numpy array of integer indices.
+        :return: An array backed `MicrographSource` object representing the micrographs for `indices`.
+        """
+        # Initialize empty result
+        n_micrographs = len(indices)
+        micrographs = np.empty(
+            (n_micrographs, self.micrograph_size, self.micrograph_size),
+            dtype=self.dtype,
+        )
+        for i, ind in enumerate(indices):
+            # Load the micrograph image from file
+            micrograph = Image.load(self.micrograph_files[ind])
+            # Assert size
+            if micrograph.resolution != self.micrograph_size:
+                raise NotImplementedError(
+                    f"Micrograph {ind} has inconsistent shape {micrograph.resolution}, expected {self.micrograph_size}."
+                )
+            # Assign to array, implicitly performs casting to dtype
+            micrographs[i] = micrograph.asnumpy()
+
+        return MicrographSource(micrographs)
+
+    def __repr__(self):
+        """
+        String representation.
+
+        :return: Returns a string description of instance.
+        """
+        return f"{self.__class__.__name__} with {self.micrograph_count} {self.dtype.name} micrographs of size {self.micrograph_size}x{self.micrograph_size}"
+
+    def __len__(self):
+        """
+        Returns the number of micrographs.
+
+        :return: Returns the micrograph count
+        """
+        return self.micrograph_count
+
+    def save(self, path):
+        """
+        Save micrographs to `path`.
+
+        Currently only support saving to `.mrc`.
+
+        :param path: Path to save data.
+        """
+        ArrayImageSource(self.asnumpy()).save(path, batch_size=1)
+
+    def _glob_files(self, file_path):
+        """
+        Helper utility to glob for matching micrograph files at `file_path`.
+
+        Will match on any image matching `Image.extensions` in `file_path`.
+
+        :return: List of files.
+        """
+
+        files = []
+        # Determine if directory or single file.
+        if os.path.isdir(file_path):
+            # Add all loadable images in the directory.
+            for ext in Image.extensions:
+                files.extend(sorted(glob(os.path.join(file_path, f"*{ext}"))))
+            # Ensure we have not failed to find any micrographs in the provided directory.
+            if len(files) == 0:
+                raise RuntimeError(f"No suitable files were found at {file_path}.")
```

Comment:
Should we test this branch? And the one below (line 199)?

Reply:
Sure why not.